crypto/sm2/sm2_crypt.c

```diff
@@ -0,0 +1,262 @@
+    EC_POINT *kG = EC_POINT_new(group);
+    EC_POINT *kP = EC_POINT_new(group);
+    const EC_POINT *P = EC_KEY_get0_public_key(key);
+    uint8_t msg_mask[msg_len];
+
+    const size_t field_size = EC_field_size(group);
+    uint8_t x2y2[2 * field_size];
+    uint8_t x2_bits[field_size];
+    uint8_t y2_bits[field_size];
+
+    const size_t C3_size = EVP_MD_size(digest);
+    uint8_t C3[C3_size];
+
+    memset(ciphertext_buf, 0, *ciphertext_len);
+
+    BN_priv_rand_range(k, order);
+
+    if (EC_POINT_mul(group, kG, k, NULL, NULL, ctx) == 0)
+        goto done;
+
+    if (EC_POINT_get_affine_coordinates_GFp(group, kG, x1, y1, ctx) == 0)
+        goto done;
+
+    if (EC_POINT_mul(group, kP, NULL, P, k, ctx) == 0)
+        goto done;
+
+    if (EC_POINT_get_affine_coordinates_GFp(group, kP, x2, y2, ctx) == 0)
+        goto done;
+
+    BN_bn2binpad(x2, x2_bits, field_size);
+    BN_bn2binpad(y2, y2_bits, field_size);
+
+    memcpy(x2y2, x2_bits, field_size);
+    memcpy(x2y2 + field_size, y2_bits, field_size);
+
+    // This happens to match the KDF used in SM2
+    if (ECDH_KDF_X9_62(msg_mask, msg_len, x2y2, 2 * field_size, NULL, 0, digest)
+        == 0)
+        goto done;
+
+    for (size_t i = 0; i != msg_len; ++i)
+        msg_mask[i] ^= msg[i];
+
+    if (EVP_DigestInit(hash, digest) == 0)
+        goto done;
+
+    if (EVP_DigestUpdate(hash, x2_bits, field_size) == 0)
+        goto done;
+
+    if (EVP_DigestUpdate(hash, msg, msg_len) == 0)
+        goto done;
+
+    if (EVP_DigestUpdate(hash, y2_bits, field_size) == 0)
+        goto done;
+
+    if (EVP_DigestFinal(hash, C3, NULL) == 0)
+        goto done;
+
+    struct SM2_Ciphertext_st ctext_struct;
+    ctext_struct.C1x = x1;
+    ctext_struct.C1y = y1;
+    ctext_struct.C3 = ASN1_OCTET_STRING_new();
+    ASN1_OCTET_STRING_set(ctext_struct.C3, C3, C3_size);
+    ctext_struct.C2 = ASN1_OCTET_STRING_new();
+    ASN1_OCTET_STRING_set(ctext_struct.C2, msg_mask, msg_len);
+
+    *ciphertext_len = i2d_SM2_Ciphertext(&ctext_struct, &ciphertext_buf);
+
+    ASN1_OCTET_STRING_free(ctext_struct.C2);
+    ASN1_OCTET_STRING_free(ctext_struct.C3);
+
+    rc = 1;
+
+ done:
+    BN_free(k);
+    BN_free(x1);
+    BN_free(y1);
+    BN_free(x2);
+    BN_free(y2);
+    EVP_MD_CTX_free(hash);
+    BN_CTX_free(ctx);
+    EC_POINT_free(kG);
+    EC_POINT_free(kP);
+    return rc;
+}
+
+int SM2_decrypt(const EC_KEY *key,
+                const EVP_MD *digest,
+                const uint8_t *ciphertext,
+                size_t ciphertext_len, uint8_t *ptext_buf, size_t *ptext_len)
+{
+    int rc = 0;
+
+    BN_CTX *ctx = BN_CTX_new();
+    const EC_GROUP *group = EC_KEY_get0_group(key);
+    EC_POINT *C1 = NULL;
+    struct SM2_Ciphertext_st *sm2_ctext = NULL;
+    const size_t field_size = EC_field_size(group);
+    uint8_t x2y2[2 * field_size];
+    uint8_t x2_bits[field_size];
+    uint8_t y2_bits[field_size];
+    BIGNUM *x2 = BN_new();
+    BIGNUM *y2 = BN_new();
```

Comment:
Same comment here about starting and allocating from the ctx for these bignums.
Again same comments in this function about checking for NULL's from allocations.